runtime/engine/src/main/java/io/aklivity/zilla/runtime/engine/Engine.java

```diff
@@ -481,6 +484,18 @@ public long[][] histogramIds()
         return worker.histogramIds();
     }
 
+    public Supplier<RingBufferSpy>[] supplyEventSpies(
+        SpyPosition position)
+    {
+        @SuppressWarnings("unchecked")
+        Supplier<RingBufferSpy>[] spies = new Supplier[workers.size()];
+        for (int i = 0; i < workers.size(); i++)
+        {
+            spies[i] = workers.get(i).supplyEventSpy(position);
+        }
+        return spies;
+    }
+
```

Comment:
Generally we try to avoid allocation on read such as this.

When we need state to span across all binding contexts for example, then it is typically created once at the binding and passed to each binding context.

The same style approach works here, instantiating the state once in the engine and passing to all engine contexts.

However, in this case the state would seem to be an array of `Function<SpyPosition, RingBufferSpy>` so that callers can create their own `RingBufferSpy` instances to walk over.

We might prefer to further abstract this, keeping the `RingBufferSpy` concept an internal implementation detail and instead just providing an `EventReader` with `int read(MessageConsumer)` method for example.

Reply:
OK